### PR TITLE
fix diskann async cache generation

### PR DESCRIPTION
**Pull request by foxspy**
issue: #372

### CODE REVIEW

File: include/knowhere/comp/thread_pool.h [hunk 1]
```diff
@@ -48,23 +51,33 @@ class ThreadPool {
...
 
  public:
-    explicit ThreadPool(uint32_t num_threads, const std::string& thread_name_prefix)
-        : pool_(folly::CPUThreadPoolExecutor(
-              num_threads,
-              std::make_unique<
-                  folly::LifoSemMPMCQueue<folly::CPUThreadPoolExecutor::CPUTask, folly::QueueBehaviorIfFull::BLOCK>>(
-                  num_threads * kTaskQueueFactor),
-              std::make_shared<LowPriorityThreadFactory>(thread_name_prefix))) {
+    explicit ThreadPool(uint32_t num_threads, const std::string& thread_name_prefix, QueueType queueT = QueueType::LIFO)
```
Comment (alexanderguzhva) [inline include/knowhere/comp/thread_pool.h:54]: Personally, I would introduce `static ThreadPool CreateFIFO()` and `static ThreadPool CreateLIFO()`. These are two very different behaviors which one does not control via a boolean flag, and it make the code more readable.

Comment (liliu-z) [reply include/knowhere/comp/thread_pool.h:54]: Sounds good

File: thirdparty/DiskANN/src/pq_flash_index.cpp [hunk 2]
```diff
@@ -268,6 +269,19 @@ namespace diskann {
     }
     this->count_visited_nodes.store(true);
 
+    // sync allocate memory
+    if (nhood_cache_buf == nullptr) {
+        nhood_cache_buf = new unsigned[num_nodes_to_cache * (max_degree + 1)];
+        memset(nhood_cache_buf, 0, num_nodes_to_cache * (max_degree + 1));
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/src/pq_flash_index.cpp:275]: `sizeof(unsigned)` is 4 on my machine, so memset should include `sizeof(unsigned)` as well.
But instead I would just  convert `nhood_cache_buf` from a pointer to a `std::vector<>`.

File: thirdparty/DiskANN/src/pq_flash_index.cpp [hunk 3]
```diff
@@ -190,18 +189,20 @@ namespace diskann {
...
-    assert(this->nhood_cache_buf == nullptr && "nhoodc_cache_buf is not null");
-    assert(this->coord_cache_buf == nullptr && "coord_cache_buf is not null");
 
     auto ctx = this->reader->get_ctx();
 
-    nhood_cache_buf = new unsigned[num_cached_nodes * (max_degree + 1)];
-    memset(nhood_cache_buf, 0, num_cached_nodes * (max_degree + 1));
+    if (nhood_cache_buf == nullptr) {
+        nhood_cache_buf = new unsigned[num_cached_nodes * (max_degree + 1)];
+        memset(nhood_cache_buf, 0, num_cached_nodes * (max_degree + 1));
```
Comment (alexanderguzhva) [inline thirdparty/DiskANN/src/pq_flash_index.cpp:197]: `sizeof(unsigned)` is 4 on my machine, so memset should include `sizeof(unsigned)` as well.
But instead I would just  convert `nhood_cache_buf` from a pointer to a `std::vector<>`.

**Review by alexanderguzhva — CHANGES_REQUESTED** [review]

**Pull request merged**